### ISSUE TITLE
The app crashes after deleting and adding a project again

### ISSUE THREAD

**Issue by grzesiek2010** [getodk/collect#5042]
#### Problem description
The app crashes after deleting and adding a project again.

```
Caused by: android.database.sqlite.SQLiteReadOnlyDatabaseException: attempt to write a readonly database (code 1032 SQLITE_READONLY_DBMOVED)
        at android.database.sqlite.SQLiteConnection.nativeExecuteForChangedRowCount(Native Method)
        at android.database.sqlite.SQLiteConnection.executeForChangedRowCount(SQLiteConnection.java:890)
        at android.database.sqlite.SQLiteSession.executeForChangedRowCount(SQLiteSession.java:756)
        at android.database.sqlite.SQLiteStatement.executeUpdateDelete(SQLiteStatement.java:66)
        at android.database.sqlite.SQLiteDatabase.delete(SQLiteDatabase.java:1729)
        at org.odk.collect.android.database.forms.DatabaseFormsRepository.deleteForms(DatabaseFormsRepository.java:239)
        at org.odk.collect.android.database.forms.DatabaseFormsRepository.delete(DatabaseFormsRepository.java:164)
        at org.odk.collect.android.utilities.FormsDirDiskFormsSynchronizer.synchronizeAndReturnError(FormsDirDiskFormsSynchronizer.java:101)
        at org.odk.collect.android.utilities.FormsDirDiskFormsSynchronizer.synchronize(FormsDirDiskFormsSynchronizer.java:44)
        at org.odk.collect.android.formmanagement.ServerFormsDetailsFetcher.fetchFormDetails(ServerFormsDetailsFetcher.kt:51)
        at org.odk.collect.android.tasks.DownloadFormListTask.doInBackground(DownloadFormListTask.java:63)
        at org.odk.collect.android.tasks.DownloadFormListTask.doInBackground(DownloadFormListTask.java:39)
```

#### Steps to reproduce the problem
1. Crete Demo project
2. Download one form
3. Delete the project
4. Create Demo project again
5. Try to download forms.

#### Other information 
The issue is similar to https://github.com/getodk/collect/issues/4870 but it looks like not the same.

**Comment by seadowg** [comment]
Reopening this seeing as we are reverting #5043. After spending an afternoon working around the problems we saw with adding `close()` in #5043 I have some thoughts:

* Our current code has no protections against "connections" (cached `SQLiteOpenHelper` instances) being closed while in use which makes adding any level of clean up hard
* Caching `SQLiteOpenHelper` objects is theoretically good for performance, but we have no actual evidence that it's a helpful
* If we moved away from caching `SQLiteOpenHelper` objects (which I think we should) we'll need to rework our code (probably in `DatabaseConnection` and the repo classes) so that `close()` is called when we're done with a particular query
* No matter what we do, it's possible for the user to create weird situations by allowing them to delete databases

**Comment by lognaturel** [comment]
I just ran into this accidentally.

This only applies to the `DEMO` project, right? Because its projectid is predictable? I think there are a lot of weird properties of that special case and we should perhaps consider removing it.

**Comment by seadowg** [comment]
> This only applies to the DEMO project, right? Because its projectid is predictable? I think there are a lot of weird properties of that special case and we should perhaps consider removing it.

I don't **think** so? I'd imagine you'd run into a similar scenario if you delete the database file(s) for any project and then return to the app as you'd still have a cached `SQLiteOpenHelper` for the deleted file. We should confirm/deny that though.

**Comment by lognaturel** [comment]
> delete the database file(s) for any project and then return to the app

I think in that case you wouldn't be recreating the same project because it would have a different project id. I quickly created a project from a qr code, deleted it, recreated with the same qr code and all of that seemed to work fine. I can systematically reproduce with the demo project.

**Comment by seadowg** [comment]
> I think in that case you wouldn't be recreating the same project because it would have a different project id.

I meant deleting the files themselves rather than deleting through Collect. In that case, the files should get recreated, so you'll end up in the same scenario: a cached `SQLiteOpenHelper` pointing at a file that's been deleted and then recreated. We should investigate and see if that causes the same problem as we see with DEMO.

**Comment by seadowg** [comment]
As an aside I keep wondering if there's a way for us to get notifications when the file system changes - I know a lot of desktop apps (media players etc) do this. I failed at Googling this usually but this morning I think I found what we need: https://developer.android.com/reference/android/os/FileObserver. I'm wondering if long term we can use that to simplify all these problems we have with the file system getting out of sync.
